### ISSUE TITLE
mono plugin - sometimes not all data gets sent to the client

### ISSUE THREAD

**Issue by Evengard** [unbit/uwsgi#1919]
This is a really weird problem which I don't know where it does come from. From time to time, when loading a page under heavy load, it just doesn't load completely. It seems like the mono runtime generates it fully, but it actually doesn't end up on client-side. I was able to reproduce the issue on kinda heavy-weight pages, with running such scripts as:

```bash
#!/bin/bash
for (( i=1; i <= 100; i++ ))
do
curl 'https://$DOMAIN/' --silent --write-out 'size_download=%{size_download}\n' --output /dev/null &
done
```

The page size of a page which was expected to have the same size differed!

I've also then rechecked with another script, such as:

```bash
#!/bin/bash
for (( i=1; i <= 100; i++ ))
do
curl 'https://$DOMAIN/' --silent | grep -o '.............$'  &
done
```

I was expecting getting the last closing tag of an html page - `</html>`, and sometimes it was... But sometimes it ended up in mid-page html code!

My app is running in multithreaded mode, that may be a kind of a cause, but still this is a bug. It seems like some buffers doesn't get sent correctly.

**Comment by Evengard** [comment]
This one is actually really weird. It seems that we're getting incomplete data... From Mono itself 😕 The sizes passed to SendResponseFromMemory matches the ones received on the client completely. I even checked the passed array sizes with mono_array_length - they also matches the length. That's kind of weird because it happens only on simultaneous requests. And that's also weird because it never happened on hyperfastcgi.

**Comment by rdeioris** [comment]
In the uwsgi logs you have the size of the response. Does it match with the generated data ?

**Comment by Evengard** [comment]
It matches with the data sent to the client. But it doesn't matches the "expected" file size all the time.

It really seems that when another thread calls `HttpRuntime.ProcessRequest` it somewhat cancels out some other (probably previous) request data sending, that's why it doesn't end up sending all the data - on the mono side, on the uwsgi side the actual sending seems to be fine... Which is weird because it really isn't supposed to do like this. Some kind of witchcraft involved - seems like that for me. Really weird.

**Comment by Evengard** [comment]
https://github.com/mono/mono/blob/2fd171c4f9d08d0f96917cb85d071071dcaa57ed/mcs/class/System.Web/System.Web/HttpRuntime.cs#L488-L489

According to this, `HttpRuntime.ProcessRequest` isn't guaranteed to be synchronous. This may be the cause... We may be loosing the references to the `current_wsgi_req()` because it may be executed on some other thread... Maybe. That's just a theory. I think it may be related to the asynchronous nature of Mono internals. And we are finishing the request on `ProcessRequest` return, which may be incorrect, as the request may still be waiting asynchronously for something. Really, I don't know, just theorizing here. I guess we might try at first to get some "EndRequest" event first and only then finishing the request.

**Comment by Evengard** [comment]
Welp, I managed to pass the pointer to `current_wsgi_req()` to the object, ditching all the calls to the actual function after the request is started, and added a lock on wait until the request is complete... Except that it didn't helped with this exact bug. I think I'll try to wrap the call to `HttpRuntime.ProcessRequest()` into a task next... Not sure what can help.

**Comment by Evengard** [comment]
It seems like Mono is messing around with its signals here. I guess we need to use ppoll here and block the "unneeded" Mono signals here so it works as intended.

**Comment by Evengard** [comment]
Yes! Calling to ppoll with blocking all the signals worked like a charm! The bug is gone. Pull request incoming soon.
